core/vm/contracts.go

```diff
@@ -1258,3 +1288,150 @@ func (c *blsSignatureVerify) Run(input []byte) ([]byte, error) {
+func (c *verifyHeader) RequiredGas(input []byte) uint64 {
+	msgAndSigLength := msgHashLength + signatureLength
+	inputLen := uint64(len(input))
+	if inputLen <= msgAndSigLength ||
+		(inputLen-msgAndSigLength)%singleBlsPubkeyLength != 0 {
+		return params.BlsSignatureVerifyBaseGas
+	}
+	pubKeyNumber := (inputLen - msgAndSigLength) / singleBlsPubkeyLength
+	return params.BlsSignatureVerifyBaseGas + pubKeyNumber*params.BlsSignatureVerifyPerKeyGas
+}
+
+var (
+	bscHeaderType, _ = abi.NewType("tuple", "", []abi.ArgumentMarshaling{
+		{Name: "ParentHash", Type: "bytes32"},
+		{Name: "UncleHash", Type: "bytes32"},
+		{Name: "Coinbase", Type: "address"},
+		{Name: "Root", Type: "bytes32"},
+		{Name: "TxHash", Type: "bytes32"},
+		{Name: "ReceiptHash", Type: "bytes32"},
+		{Name: "Bloom", Type: "bytes"},
+		{Name: "Difficulty", Type: "uint64"},
+		{Name: "Number", Type: "uint64"},
+		{Name: "GasLimit", Type: "uint64"},
+		{Name: "GasUsed", Type: "uint64"},
+		{Name: "Time", Type: "uint64"},
+		{Name: "Extra", Type: "bytes"},
+		{Name: "MixDigest", Type: "bytes32"},
+		{Name: "Nonce", Type: "bytes8"},
+	})
+
+	bscHeaderArgs = abi.Arguments{
+		{Type: bscHeaderType},
+	}
+
+	extraSeal = 65
+)
+
+type BscHeaderStruct struct {
+	ParentHash  [32]byte
+	UncleHash   [32]byte
+	Coinbase    common.Address
+	Root        [32]byte
+	TxHash      [32]byte
+	ReceiptHash [32]byte
+	Bloom       []byte
+	Difficulty  uint64
+	Number      uint64
+	GasLimit    uint64
+	GasUsed     uint64
+	Time        uint64
+	Extra       []byte
+	MixDigest   [32]byte
+	Nonce       [8]byte
+}
+
+func (c *verifyHeader) Run(input []byte) ([]byte, error) {
+	if len(input) < 32 {
+		log.Debug("verifyHeader wrong input size", "inputLen", len(input))
+		return nil, ErrExecutionReverted
+	}
+
+	// get chainId
+	chainId := new(big.Int).SetBytes(input[:32])
+
+	// get header
+	bscHeaderBz := getData(input, 32, uint64(len(input)-32))
+	unpacked, err := bscHeaderArgs.Unpack(bscHeaderBz)
+	if err != nil {
+		log.Debug("verifyHeader invalid header", "err", err)
+		return nil, ErrExecutionReverted
+	}
+	unpackedStruct := abi.ConvertType(unpacked[0], BscHeaderStruct{})
+	bscHeader, ok := unpackedStruct.(BscHeaderStruct)
+	if !ok {
+		log.Debug("verifyHeader invalid header", "err", err)
+		return nil, ErrExecutionReverted
+	}
+
+	// get signature
+	if len(bscHeader.Extra) < extraSeal {
+		log.Debug("verifyHeader invalid header extra", "err", err)
+		return nil, ErrExecutionReverted
+	}
+	signature := bscHeader.Extra[len(bscHeader.Extra)-extraSeal:]
+
+	// verify signature
+	signHash, err := sealHash(&bscHeader, chainId)
+	if err != nil {
+		log.Debug("verifyHeader seal hash failed", "err", err)
+		return nil, ErrExecutionReverted
+	}
+	pubKey, err := secp256k1.RecoverPubkey(signHash.Bytes(), signature)
+	if err != nil {
+		log.Debug("verifyHeader recover pubkey failed", "err", err)
+		return nil, ErrExecutionReverted
+	}
+
+	return crypto.Keccak256(pubKey[1:])[12:], nil
```

Comment:
I think this function only return the signer of the block header.  For a double sign slash, the logic is more complicated: like compare two header, they should signed by same validator and at the same height, with same parent block. 

It cost a lot of cast to implement this in contract, why dont we implement that in this precompile comtract?

Reply:
Done